File: bokeh/embed/notebook/widgets.py
#-----------------------------------------------------------------------------
# Copyright (c) 2012 - 2019, Anaconda, Inc., and Bokeh Contributors.
# All rights reserved.
#
# The full license is in the file LICENSE.txt, distributed with this software.
#-----------------------------------------------------------------------------
"""
Defines a Bokeh model wrapper for Jupyter notebook/lab, which renders Bokeh models
and performs bi-directional syncing just like bokeh server does.

"""

#-----------------------------------------------------------------------------
# Boilerplate
#-----------------------------------------------------------------------------

#-----------------------------------------------------------------------------
# Imports
#-----------------------------------------------------------------------------

# Standard library imports
import json

# External imports

# Bokeh imports
from ...core.json_encoder import serialize_json
from ...models import LayoutDOM
from ...document import Document
from ...protocol import Protocol
from ...util.version import __version__
from ...util.dependencies import import_optional
from ..elements import div_for_render_item
from ..util import standalone_docs_json_and_render_items

#-----------------------------------------------------------------------------
# Globals and constants
#-----------------------------------------------------------------------------

__all__ = (
    "BokehModel",
)

_module_name = "bokehjs"
_module_version = __version__

ipywidgets = import_optional("ipywidgets")
traitlets = import_optional("traitlets")

Unicode = traitlets.Unicode
Dict = traitlets.Dict

#-----------------------------------------------------------------------------
# General API
#-----------------------------------------------------------------------------

class BokehModel(ipywidgets.DOMWidget):

    _model_name = Unicode("BokehModel").tag(sync=True)
    _model_module = Unicode(_module_name).tag(sync=True)
    _model_module_version = Unicode(_module_version).tag(sync=True)
    _view_name = Unicode("BokehView").tag(sync=True)
    _view_module = Unicode(_module_name).tag(sync=True)
    _view_module_version = Unicode(_module_version).tag(sync=True)

    render_bundle = Dict().tag(sync=True, to_json=lambda obj, _: serialize_json(obj))

    def __init__(self, model, document=None, **kwargs):
        if isinstance(model, LayoutDOM):
            self.update_from_model(model, document)
        super(BokehModel, self).__init__(**kwargs)
        self.on_msg(self._sync_model)

    def close(self):
        if self._document is not None:
            self._document.remove_on_change(self)

    @classmethod
    def _model_to_traits(cls, model, document=None):
        if document is None:
            document = Document()
            document.add_root(model)
        (docs_json, [render_item]) = standalone_docs_json_and_render_items([model], True)
        render_bundle = dict(
            docs_json=docs_json,
            render_items=[render_item.to_json()],
            div=div_for_render_item(render_item),
        )
        return render_bundle, document

    def update_from_model(self, model, document=None):
        self._model = model
        self.render_bundle, self._document = self._model_to_traits(model, document)
        self._document.on_change_dispatch_to(self)

    def _document_patched(self, event):
        msg = Protocol("1.0").create("PATCH-DOC", [event])

        self.send({"msg": "patch", "payload": msg.header_json})
        self.send({"msg": "patch", "payload": msg.metadata_json})
        self.send({"msg": "patch", "payload": msg.content_json})
        for header, buff in msg.buffers:
            self.send({"msg": "patch", "payload": json.dumps(header)})
            self.send({"msg": "patch", "payload": buff})

    def _sync_model(self, _, content, _buffers):
        if content.get("event", "") != "jsevent":
            return
        new, old, attr = content["new"], content["old"], content["attr"]
        submodel = self._model.select_one({"id": content["id"]})
        try:
            setattr(submodel, attr, new)
        except Exception:
            return
        for cb in submodel._callbacks.get(attr, []):
            cb(attr, old, new)

#-----------------------------------------------------------------------------
# Dev API
#-----------------------------------------------------------------------------

File: bokehjs/src/lib/embed/notebook/widgets.ts
import {DOMWidgetModel, DOMWidgetView, ISerializers} from "@jupyter-widgets/base"
import {version} from "version"

import {Document, DocumentChangedEvent, ModelChangedEvent} from "document"
import {Receiver, Fragment} from "protocol/receiver"
import {keys, values} from "core/util/object"

import {add_document_standalone} from "../standalone"
import {DocsJson, RenderItem} from "../json"

export class BokehModel extends DOMWidgetModel {
  defaults(): object {
    return {
      ...super.defaults(),
      _model_name: BokehModel.model_name,
      _model_module: BokehModel.model_module,
      _model_module_version: BokehModel.model_module_version,
      _view_name: BokehModel.view_name,
      _view_module: BokehModel.view_module,
      _view_module_version: BokehModel.view_module_version,
      render_bundle: {},
    }
  }

  static serializers: ISerializers = {
    ...DOMWidgetModel.serializers,
  }

  static model_name = "BokehModel"
  static model_module = "bokehjs"
  static model_module_version = version
  static view_name = "BokehView"
  static view_module = "bokehjs"
  static view_module_version = version
}

export type RenderBundle = {
  docs_json: DocsJson
  render_items: RenderItem[]
  div: string
}

export class BokehView extends DOMWidgetView {
  private _document: Document | null
  private _receiver: Receiver
  private _blocked: boolean

  constructor(options?: any) {
    super(options)
    this._document = null
    this._blocked = false
    this._receiver = new Receiver()
    this.model.on("change:render_bundle", () => this.render())
    this.listenTo(this.model, "msg:custom", (msg) => this._consume_patch(msg))
  }

  render(): void {
    const bundle = JSON.parse(this.model.get("render_bundle"))
    const {docs_json, render_items, div} = bundle as RenderBundle
    this.el.innerHTML = div
    const element = this.el.children[0]
    const json = values(docs_json)[0]
    this._document = Document.from_json(json)
    for (const item of render_items) {
      const roots: {[key: string]: HTMLElement} = {}
      for (const root_id in item.roots)
        roots[root_id] = element
      add_document_standalone(this._document, element, roots)
    }
    this._document.on_change((event) => this._change_event(event))
  }

  protected _change_event(event: DocumentChangedEvent): void {
    if (!this._blocked && event instanceof ModelChangedEvent)
      this.send({event: "jsevent", id: event.model.id, new: event.new_, attr: event.attr, old: event.old})
  }

  protected _consume_patch(content: {msg: "patch", payload: Fragment}): void {
    if (this._document == null)
      return
    if (content.msg == "patch") {
      this._receiver.consume(content.payload)
      const comm_msg = this._receiver.message
      if (comm_msg != null && keys(comm_msg.content).length > 0) {
        this._blocked = true
        try {
          this._document.apply_json_patch(comm_msg.content, comm_msg.buffers)
        } finally {
          this._blocked = false
        }
      }
    }
  }
}
